sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -519,7 +520,9 @@ case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with
 
   override def sql: String = {
     val valueSQL = child.sql
-    val listSQL = hset.toSeq.map(Literal(_).sql).mkString(", ")
+    val listSQL = hset.toSeq
+      .map(elem => Literal(convertToScala(elem, child.dataType)).sql)
```

Comment:
this converts the internal value to external value, and then `Literal.apply` converts external value to internal value.

Can we just do `Literal(elem, child.dataType).sql`?

Reply:
@cloud-fan No, Literal fails on UTF8String value, see https://github.com/apache/spark/pull/28328#discussion_r414566719

Reply:
Another problem is https://github.com/apache/spark/pull/28328#discussion_r415256811

Reply:
So, the problem is Literal requires external types but elem has internal Catalyst's type